File: app/src/main/java/nota/android/crash/CrashHandler.java
package nota.android.crash;

import android.os.Handler;
import android.os.Looper;

import de.robv.android.xposed.XposedHelpers;

public class CrashHandler {


    public static final String SOURCE_LOOPER = "looper";
    public static final String SOURCE_UNCAUGHT = "uncaught";

    public interface ExceptionHandler {

        void handlerException(Throwable throwable, String source);
    }

    private CrashHandler() {
    }

    private static ExceptionHandler sExceptionHandler;
    private static Thread.UncaughtExceptionHandler sUncaughtExceptionHandler;
    private static boolean sInstalled = false;


    public static synchronized void insert(ExceptionHandler exceptionHandler) {
        if (sInstalled) {
            return;
        }
        sInstalled = true;
        sExceptionHandler = exceptionHandler;

        final Looper targetLooper = (Looper) XposedHelpers.callStaticMethod(Looper.class, "getMainLooper");
        new Handler(targetLooper).post(() -> {
            //main 方法的退出就是Looper.loop执行完毕，所有事件都是Looper的监听，
            //当Crash的时候，说明App的线程已经无法loop了，这时候需要这段代码给APP续命
            while (true) {
                try {
                    XposedHelpers.callStaticMethod(Looper.class, "loop");
                } catch (Throwable e) {
                   /* if (e instanceof RuntimeException) {
                        return;
                    }*/
                    if (sExceptionHandler != null) {
                        sExceptionHandler.handlerException(e, SOURCE_LOOPER);
                    }
                }
            }
        });


        sUncaughtExceptionHandler = (Thread.UncaughtExceptionHandler) XposedHelpers.callStaticMethod(Thread.class, "getDefaultUncaughtExceptionHandler");
        //原来的不给他处理哦
        XposedHelpers.callStaticMethod(Thread.class, "setDefaultUncaughtExceptionHandler", (Thread.UncaughtExceptionHandler) (t, e) -> {
            if (sExceptionHandler != null) {
                sExceptionHandler.handlerException(e, SOURCE_UNCAUGHT);
            }
        });

    }


}

File: app/src/main/java/nota/android/crash/CrashHandler.kt
package nota.android.crash

import android.os.Handler
import android.os.Looper
import android.os.Process
import de.robv.android.xposed.XposedHelpers

/**
 * Hook-side crash capture (ADR-011, ADR-023).
 *
 * [Mode.INTERCEPT]: guarded infinite Looper loop + non-forwarding UEH (swallow).
 * [Mode.OBSERVE]: one-shot Looper catch + forwarding UEH (log then system exit).
 */
object CrashHandler {

    const val SOURCE_LOOPER = "looper"
    const val SOURCE_UNCAUGHT = "uncaught"

    enum class Mode {
        INTERCEPT,
        OBSERVE,
    }

    fun interface ExceptionHandler {
        fun handleException(throwable: Throwable, source: String)
    }

    @Volatile
    private var installed = false

    @Volatile
    private var exceptionHandler: ExceptionHandler? = null

    @Volatile
    private var previousUncaughtExceptionHandler: Thread.UncaughtExceptionHandler? = null

    @JvmStatic
    @Synchronized
    fun install(mode: Mode, handler: ExceptionHandler) {
        if (installed) return
        installed = true
        exceptionHandler = handler

        val targetLooper = XposedHelpers.callStaticMethod(Looper::class.java, "getMainLooper") as Looper
        when (mode) {
            Mode.INTERCEPT -> installInterceptLoop(targetLooper)
            Mode.OBSERVE -> installObserveLoop(targetLooper)
        }
        installUncaughtHandler(mode)
    }

    private fun installInterceptLoop(targetLooper: Looper) {
        fun loopOnce() {
            try {
                XposedHelpers.callStaticMethod(Looper::class.java, "loop")
            } catch (e: Throwable) {
                exceptionHandler?.handleException(e, SOURCE_LOOPER)
                Handler(targetLooper).post(::loopOnce)
            }
        }
        Handler(targetLooper).post(::loopOnce)
    }

    private fun installObserveLoop(targetLooper: Looper) {
        fun loopOnce() {
            try {
                XposedHelpers.callStaticMethod(Looper::class.java, "loop")
            } catch (e: Throwable) {
                exceptionHandler?.handleException(e, SOURCE_LOOPER)
                throw e
            }
        }
        Handler(targetLooper).post(::loopOnce)
    }

    private fun installUncaughtHandler(mode: Mode) {
        previousUncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()
        val delegate = Thread.UncaughtExceptionHandler { thread, throwable ->
            exceptionHandler?.handleException(throwable, SOURCE_UNCAUGHT)
            when (mode) {
                Mode.INTERCEPT -> Unit
                Mode.OBSERVE -> {
                    val previous = previousUncaughtExceptionHandler
                    if (previous != null) {
                        previous.uncaughtException(thread, throwable)
                    } else {
                        Process.killProcess(Process.myPid())
                    }
                }
            }
        }
        XposedHelpers.callStaticMethod(
            Thread::class.java,
            "setDefaultUncaughtExceptionHandler",
            delegate,
        )
    }
}
